[decision] Changetracking (round 3) (#4841)
ElektraInitiative/libelektra: doc/decisions/3_in_review/change_tracking.md

- It is possible to do change tracking with reasonable memory and computation overhead
- It is possible to design a single change tracking API that is useful for all existing and future plugins
- False positives are okay

>   - this may happened when some keys have been changed by the user, but have subsequentially been "unchanged" by a transformation plugin

Reviewer: ```suggestion
  - this may happens when some keys have been changed by the user, but have subsequentially been "unchanged" by a transformation plugin
```

### Provide an API within `libelektra-kdb`

> The API should be useable both by plugins and applications utilizing ELektra.

Reviewer: ```suggestion
The API should be usable by plugins and by applications utilizing Elektra.
```

## Decision

> Plugins and application developer can declare that changetracking is required via a contract.

Reviewer: ```suggestion
Plugins and application developers can declare that changetracking is required via a contract.
```

> Plugin and application developers can declare that changetracking is required via a contract.

Reviewer 2: I don't know if it necessary to enable it via application developers. Please link a use case for this.

What definitely is needed, is that it can be enabled via tooling (and then stays enabled for everyone).

Store deep-duped copy-on-write returned keys in a separate keyset, which we might also use as [internal cache](../4_decided/internal_cache.md).
The whole changetracking logic lives within `libelektra-kdb`.
We provide an API for developers with `libelektra-kdb`.

## Rationale

This decision meets all the constraints.

We are not altering the behavior of already existing functions, so everything is transparent for applications using the public Elektra API.
As changetracking is an opt-in feature, it will create zero runtime overhead and negligible memory overhead in usecases where it is not needed.
There is only a single implementation and storage, so no duplication for each plugin that wants change tracking.
As we are storing our own tracking data, all sequences of `kdbGet` and `kdbSet` should work.

## Implications

- Documentation for changetracking APIs needs to be written
- `struct _KDB` needs to be extended to contain the keyset for changetracking
- Replace all custom changetracking code within Elektra with this unified implementation.
  This should also include the `keyNeedsSync` flag.

>   It also includes switching all notification plugins to use the new API.

Reviewer 2: Please create Issues and link them here if there is something still open.
